### ISSUE TITLE
PartDesign: Add Common (intersection) operations

### ISSUE THREAD

**Issue by FreeCAD-Bug-Importer** [FreeCAD/FreeCAD#5991]
Issue imported from https://tracker.freecad.org/view.php?id=4135

* **Reporter:** townkat
* **Date submitted:** 9/17/2019
* **FreeCAD version:** 
* **Category:** Feature
* **Status:** new
* **Tags:** Part Design

# Original report text

Please add Common (intersection) operations in PartDesign in addition to Additive and Subtractive features.

> Boolean operations in Part WB are Fusion(union) (exposed as Additive features in PartDesign), Cut (exposed as Substractive features) and Common(intersection) which is not available as features in PartDesign.

More info in the following topics:
https://forum.freecadweb.org/viewtopic.php?f=8&t=35755
https://forum.freecadweb.org/viewtopic.php?f=8&t=39297

Thank you.

> Anyone care to comment why isn't there a native PartDesign common feature? I recently designed a pill box and had to add two pockets "thorough all" to remove the extra helixes from both ends. It would have been a lot simple to just define a cylinder and trim everything that is outside of it. 

Before:
![image](https://github.com/FreeCAD/FreeCAD/assets/4140247/04229344-5442-473a-9f25-c72d34e184ee)
After:  
![image](https://github.com/FreeCAD/FreeCAD/assets/4140247/45fe3c0f-3706-4194-bcb6-c122fdca161d)

# Other bug information

* **Priority:** high
* **Severity:** feature
* **Category:** Feature
* **Updated:** 2/6/2021

**Comment by luzpaz** [comment]
@FlachyJoe is this a duplicate ?

**Comment by FlachyJoe** [comment]
@luzpaz I don't know if it's shown elsewhere but I see SubstractiveHelix has an Outside property which do this. It seems quite easy to implement for the other subtractive features.

**Comment by luzpaz** [comment]
> seems quite easy to implement

@FlachyJoe go for it ? or perhaps if you don't have time, can you elaborate here how a dev would go about adding it ?

**Comment by FlachyJoe** [comment]
The implementation for SubstractiveHelix should be duplicate for all features (maybe a common util function can be write)
Ref: https://github.com/FreeCAD/FreeCAD/blob/66ece1c0684dd6b9757a2c39a5479713555f8897/src/Mod/PartDesign/App/FeatureHelix.cpp#L344-L356

@bgbsww are you interested?

**Comment by bgbsww** [comment]
@FlachyJoe I'll take a look.

**Comment by bgbsww** [comment]
TL;DR assign me please.

So, the App side of this is pretty trivial, as @FlachyJoe identified above.  The Gui side is not as straightforward, since there is no Giu/TaskAddSubParameters in the inheritance, which means either a bunch of ugly cut and paste code in each of the six subtractive types to be added, or adding to the inheritance tree and refactoring a bit.

Also, the Helix code seems to be clearly newer and substantially different than the older AddSub subtypes, which is another place to consider refactoring to consistency.

I have a proof of concept working just fine for pocket so I'm confident, there are just a lot of moving pieces here to make it clean.

If anyone sees something I'm missing, I'm all ears.

**Comment by NomAnor** [comment]
I have an idea about refactoring the PD features, this would also allow this to be implemented: https://forum.freecad.org/viewtopic.php?t=85967

**Comment by maxwxyz** [comment]
@bgbsww could you also look into https://github.com/FreeCAD/FreeCAD/issues/11209 for this fix https://github.com/FreeCAD/FreeCAD/pull/11580

**Comment by maxwxyz** [comment]
@NomAnor are you still open for working on this?

**Comment by NomAnor** [comment]
In principle yes. I could see this as part of a bigger refactoring of PD. Maybe even some unifaction with Part. But I lack the vision how to tackle it.

**Comment by ant1ucem** [comment]
> But I lack the vision how to tackle it.

A very important feature indeed — it would save a lot of operations. I suggest implementing it in the following way:

1. Merge the pairs of buttons like Pad/Pocket, Revolution/Groove, Additive Pipe/Subtractive Pipe, and so on.
2. Add a toggle switch at the top of the task panel to select between Add, Subtract, and Common operations.

![Image](https://github.com/user-attachments/assets/19b78744-2b6b-45e0-ac27-66131ae81b3b)
